File: advent-of-code-2019/05/main.py

import os 


def loadDataFile():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    with open(os.path.join(dir_path, "data.txt"), "r") as f:
        data = f.readline()
        return [int(i) for i in data.split(",")]
    return []

def loadData():
    a = [int(i) for i in "1,0,0,3,1,1,2,3,1,3,4,3,1,5,0,3,2,1,13,19,1,9,19,23,1,6,23,27,2,27,9,31,2,6,31,35,1,5,35,39,1,10,39,43,1,43,13,47,1,47,9,51,1,51,9,55,1,55,9,59,2,9,59,63,2,9,63,67,1,5,67,71,2,13,71,75,1,6,75,79,1,10,79,83,2,6,83,87,1,87,5,91,1,91,9,95,1,95,10,99,2,9,99,103,1,5,103,107,1,5,107,111,2,111,10,115,1,6,115,119,2,10,119,123,1,6,123,127,1,127,5,131,2,9,131,135,1,5,135,139,1,139,10,143,1,143,2,147,1,147,5,0,99,2,0,14,0".split(",")]
    a[1] = 12
    a[2] = 2
    return a


def loadSample():
    return [int(i) for i in "1,9,10,3,2,3,11,0,99,30,40,50".split(",")]

def loadTest():
    return [int(i) for i in "1,0,0,3,99".split(",")]

def loadStr(s):
    return [int(i) for i in s.split(",")]

Input = 1

def get_val(s, p, immediate):
    if immediate:
        return s[p]
    return s[s[p]]

def exec_op(s, op, p, immediate=[False, False]):
    if op == 1:
        s[s[p+3]] = get_val(s, p+1, immediate[0]) + get_val(s, p+2, immediate[1])
        p += 4
    elif op == 2:
        s[s[p+3]] = get_val(s, p+1, immediate[0]) * get_val(s, p+2, immediate[1])
        p += 4
    elif op == 3:
        s[s[p+1]] = Input
        p += 2
    elif op == 4:
        print(s[s[p+1]])
        # if immediate[0]:
        #     print(s[s[p+1]])
        # else:
        #     print(s[p+1])
        p += 2
    elif op == 99:
        return False
    elif op > 99:
        val = str(op).zfill(5)
        code_op = int(val[-2:])
        p = exec_op(s, code_op, p, immediate=[val[-3] == "1", val[-4] == "1", val[-5] == "1"])
    return p


def process(loadData):
    s = loadData()
    p = 0
    c = True
    print("")
    print("---")
    print("process:")
    while p is not False:
        p = exec_op(s, s[p], p)
    print("final:" + ",".join([str(i) for i in s]))

def main():
    process(lambda: loadStr("3,0,4,0,99"))
    process(lambda: loadStr("1002,4,3,4,33"))
    process(lambda: loadStr("1101,100,-1,4,0"))
    process(loadDataFile)

main()

File: advent-of-code-2019/05/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func parseData(data string) []int {
	ds := strings.Split(data, ",")
	r := []int{}
	for _, d := range ds {
		i, _ := strconv.Atoi(d)
		r = append(r, i)
	}
	return r
}

func parseFile(filename string) []int {
	file, _ := os.Open(filename)
	fscanner := bufio.NewScanner(file)
	fscanner.Scan()
	return parseData(fscanner.Text())
}

type Amp struct {
	Name      string
	Phase     int
	Code      []int
	Pos       int
	Input     chan int
	Output    chan int
	IsRunning bool
}

func (a *Amp) Start() {
	a.IsRunning = true
	go func() {
		for a.IsRunning {
			a.ExecOp()
		}
	}()
	a.Input <- a.Phase
}

func (a *Amp) GetVal(p int, immediate bool) int {
	if immediate {
		return a.Code[p]
	}
	return a.Code[a.Code[p]]
}

func getImmediate(i int, immediates []bool) bool {
	if immediates == nil {
		return false
	}
	return immediates[i]
}

func (a *Amp) ExecOp() {
	op := a.Code[a.Pos]

	var immediates []bool
	if op > 99 {
		val := fmt.Sprintf("%05d", op)
		immediates = []bool{
			val[2] == '1',
			val[1] == '1',
		}
		op = op % 100
	}

	switch op {
	case 1:
		a.Code[a.Code[a.Pos+3]] = a.GetVal(a.Pos+1, getImmediate(0, immediates)) + a.GetVal(a.Pos+2, getImmediate(1, immediates))
		a.Pos += 4
	case 2:
		a.Code[a.Code[a.Pos+3]] = a.GetVal(a.Pos+1, getImmediate(0, immediates)) * a.GetVal(a.Pos+2, getImmediate(1, immediates))
		a.Pos += 4
	case 3:
		a.Code[a.Code[a.Pos+1]] = <-a.Input
		// fmt.Printf("\t amp %s read value %d\n", a.Name, a.Code[a.Pos+1])
		a.Pos += 2
	case 4:
		// fmt.Printf("\t amp %s write value %d\n", a.Name, a.Code[a.Pos+1])
		a.Output <- a.Code[a.Code[a.Pos+1]]
		a.Pos += 2
	case 5:
		if a.GetVal(a.Pos+1, getImmediate(0, immediates)) != 0 {
			a.Pos = a.GetVal(a.Pos+2, getImmediate(1, immediates))
		} else {
			a.Pos += 3
		}
	case 6:
		if a.GetVal(a.Pos+1, getImmediate(0, immediates)) == 0 {
			a.Pos = a.GetVal(a.Pos+2, getImmediate(1, immediates))
		} else {
			a.Pos += 3
		}
	case 7:
		if a.GetVal(a.Pos+1, getImmediate(0, immediates)) < a.GetVal(a.Pos+2, getImmediate(1, immediates)) {
			a.Code[a.Code[a.Pos+3]] = 1
		} else {
			a.Code[a.Code[a.Pos+3]] = 0
		}
		a.Pos += 4
	case 8:
		if a.GetVal(a.Pos+1, getImmediate(0, immediates)) == a.GetVal(a.Pos+2, getImmediate(1, immediates)) {
			a.Code[a.Code[a.Pos+3]] = 1
		} else {
			a.Code[a.Code[a.Pos+3]] = 0
		}
		a.Pos += 4
	case 99:
		fmt.Printf("\t amp %s is stopping\n", a.Name)
		a.IsRunning = false
	default:
		panic(fmt.Sprintf("unknow op: %d", op))	
	}
}


func getAmp(code []int) *Amp {
	cInput := make(chan int)
	cOutput := make(chan int)
	return &Amp{
		Name:   "A",
		Phase:  1,
		Code:   append(code[:0:0], code...),
		Pos:    0,
		Input:  cInput,
		Output: cOutput,
	}
}

func processCode(code []int) {
	fmt.Println("---")
	fmt.Println("process")
	a := getAmp(code)
	a.Start()
	for a.IsRunning {
		select {
		case o := <-a.Output:
			fmt.Println(o)
		default:
		}
	}
	close(a.Input)
	close(a.Output)
}

func main() {
	processCode(parseData("3,9,8,9,10,9,4,9,99,-1,8"))
	processCode(parseData("3,3,1105,-1,9,1101,0,0,12,4,12,99,1"))
	processCode(parseFile("data.txt"))
	// processCode(parseData("3,23,3,24,1002,24,10,24,1002,23,-1,23,101,5,23,23,1,24,23,23,4,23,99,0,0"))
	// processCode(parseData("3,31,3,32,1002,32,10,32,1001,31,-2,31,1007,31,0,33,1002,33,7,33,1,33,31,31,1,32,31,31,4,31,99,0,0,0"))
}
